File: solution/992. Subarrays with K Different Integers/992.cpp
class Solution {
public:
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        return atMost(nums, k) - atMost(nums, k - 1);
    }
    
private:
    int atMost(vector<int>& nums, int k) {
        int res = 0;
        unordered_map<int, int> m;
        for (int i = 0, j = 0; j < nums.size(); ++j) {
            if (m[nums[j]]++ == 0) --k;
            while (k < 0) {
                if (--m[nums[i++]] == 0) ++k;
            }
            res += j - i + 1;
        }
        return res;
    }
};

File: solution/992. Subarrays with K Different Integers/992.py
class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        def atMost(n):
            count = Counter()
            res = i = 0
            for j in range(len(nums)):
                if count[nums[j]] == 0:
                    n -= 1
                count[nums[j]] += 1
                while n < 0:
                    count[nums[i]] -= 1
                    if count[nums[i]] == 0:
                        n += 1
                    i += 1
                res += j - i + 1
            return res
        
        return atMost(k) - atMost(k - 1)
    
